File: X_API_Master/X.Util.Core/Kernel/ChannelContext.cs
using System;
using X.Util.Entities;
using X.Util.Entities.Interface;

namespace X.Util.Core.Kernel
{
    public class ChannelContext<TResult, TChannel> : IContext<TResult, TChannel>
    {
        private readonly IProvider<TChannel> _channel;

        public ChannelContext(IProvider<TChannel> channel)
        {
            _channel = channel;
        }

        public IProvider<TChannel> Channel
        {
            get { return _channel; }
        }

        public Func<TResult> Calling(ActionContext<TResult> context, Func<TResult> caller)
        {
            return caller;
        }

        public void Called(ActionContext<TResult> context)
        {
            var channel = Channel as IClientProvider<TChannel>;
            if (channel != null) channel.ReleaseClient();
        }

        public void OnException(ActionContext<TResult> context)
        {
            var channel = Channel as IClientProvider<TChannel>;
            if (channel == null) return;
            channel.Dispose(channel.Domain);
        }

        public int Priority { get { return int.MaxValue; } }
    }

    public class ChannelContext<TChannel> : IContext<TChannel>
    {
        private readonly IProvider<TChannel> _channel;

        public ChannelContext(IProvider<TChannel> channel)
        {
            _channel = channel;
        }

        public IProvider<TChannel> Channel
        {
            get { return _channel; }
        }

        public Action Calling(ActionContext context, Action caller)
        {
            return caller;
        }

        public void Called(ActionContext context)
        {
            var channel = Channel as IClientProvider<TChannel>;
            if (channel != null) channel.ReleaseClient();
        }

        public void OnException(ActionContext context)
        {
            var channel = Channel as IClientProvider<TChannel>;
            if (channel == null) return;
            channel.Dispose(channel.Domain);
        }

        public int Priority { get { return int.MaxValue; } }
    }
}

File: X_Front_Master/x.front.type-components/src/util/core/channelContext.js
"use strict";
import { IProvider } from "./provider.js";

function ChannelContext(Provider) {
    this.Provider = Provider;
}

/**
 * Calling
 * @param {any} context:Context.getActionContext
 * @param {any} caller
 */
ChannelContext.prototype.Calling = function(context, caller) {
    return caller;
}

/**
 * Called
 * @param {any} context:Context.getActionContext
 */
ChannelContext.prototype.Called = function(context) {
    console.debug("debug:ChannelContext-Called;ignore", context);
    if (this.Provider instanceof IProvider.ClientProvider) {
        this.Provider.ReleaseClient();
    }
}

/**
 * OnException
 * @param {any} context:Context.getActionContext
 */
ChannelContext.prototype.OnException = function(context) {
    console.debug("debug:ChannelContext-OnException;ignore", context);
    if (this.Provider instanceof IProvider.ClientProvider) {
        this.Provider.Dispose();
    }
}

export default ChannelContext;